monix-tail/shared/src/main/scala/monix/tail/internal/IterantDropLast.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import cats.effect.Sync
+import cats.syntax.all._
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.batches.{Batch, BatchCursor}
+
+import scala.collection.mutable
+
+private[tail] object IterantDropLast {
+  /**
+    * Implementation for `Iterant#dropLast`
+    */
+  def apply[F[_], A](source: Iterant[F, A], n: Int)(implicit F: Sync[F]): Iterant[F, A] = {
+
+    def processCursor(toDrop: Int, length: Int, queue: mutable.Queue[A],
```

Comment:
Using a mutable data structure, if properly encapsulated, is fine in Monix's codebase 😀

However I try to avoid optimizations without having some proof that they work. Because if say this optimization here only brings marginal improvements, then we shouldn't do it because an immutable `Queue` is more reasonable.

So you can either revert to an immutable `Queue` for now, or make a JMH test in `benchmarks` that proves a mutable queue is better for a stream of 100000 events where you're dropping the last 1, 10 or 50 items lets say.

Don't commit the benchmark, just something to report back that indeed this is worth it.

Reply:
I generally prefer to follow the same rule but I feel like operating on ```immutable.Queue``` is pretty weird, although it might be better if there is implementation similar to how ```State``` works. 

Below are results from benchmarks but please tell me if I did some benchmarking "faux pas", I know there are a lot of catches which make results false.

Results:
``` scala
jmh:run -i 50 -wi 10 -f 2 -t 1 monix.benchmarks.IterantDropLastBenchmark

mutable.Queue
[info] Benchmark                         Mode  Cnt   Score    Error  Units
[info] IterantDropLastBenchmark.drop1   thrpt  100  73,519 â–’ 4,664  ops/s
[info] IterantDropLastBenchmark.drop10  thrpt  100  79,834 â–’ 1,388  ops/s
[info] IterantDropLastBenchmark.drop50  thrpt  100  76,795 â–’ 3,076  ops/s

immutable.Queue
[info] Benchmark                         Mode  Cnt    Score   Error  Units
[info] IterantDropLastBenchmark.drop1   thrpt  100  113,273 â–’ 2,272  ops/s
[info] IterantDropLastBenchmark.drop10  thrpt  100  117,181 â–’ 1,738  ops/s
[info] IterantDropLastBenchmark.drop50  thrpt  100  119,765 â–’ 1,839  ops/s
```
Assuming those results are true... well that's a good lesson :D I was sure that all those extra tuple allocations in ```immutable.Queue``` would show in performance. I guess ```immutable.Queue``` is better implemented?

Benchmark code:
```scala
@State(Scope.Thread)
@BenchmarkMode(Array(Mode.Throughput))
@OutputTimeUnit(TimeUnit.SECONDS)
class IterantDropLastBenchmark {

  @Benchmark
  def drop1(): List[Int] = IterantDropLastBenchmark.test(100000, 1)

  @Benchmark
  def drop10(): List[Int] = IterantDropLastBenchmark.test(100000, 1)

  @Benchmark
  def drop50(): List[Int] = IterantDropLastBenchmark.test(100000, 1)

}

object IterantDropLastBenchmark {
  def test(size: Int, toDrop: Int): List[Int] = {
    val list = List.range(1, size)
    val stream = Iterant[Task].fromList(list)

    Await.result(stream.dropLast(toDrop).toListL.runAsync, Duration.Inf)
  }
}
```
This is code for ```Iterant#dropLast``` using ```immutable.Queue``` (I have left it as a mutable-immutable hybrid monster to make this comparison more fair):
```scala
package monix.tail.internal

import cats.effect.Sync
import cats.syntax.all._
import monix.execution.misc.NonFatal
import monix.tail.Iterant
import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
import monix.tail.batches.{Batch, BatchCursor}

import scala.collection.immutable.Queue
import scala.collection.mutable

private[tail] object IterantDropLast {
  /**
    * Implementation for `Iterant#dropLast`
    */
  def apply[F[_], A](source: Iterant[F, A], n: Int)(implicit F: Sync[F]): Iterant[F, A] = {

    def processCursor(toDrop: Int, length: Int, queue: Queue[A],
                      ref: NextCursor[F, A]): Iterant[F, A] = {
      val NextCursor(cursor, rest, stop) = ref
      val limit = cursor.recommendedBatchSize
      val buffer = mutable.Buffer[A]()
      var queueLength = length

      @tailrec
      def queueLoop(queue: Queue[A], toProcess: Int): Queue[A] = {
        if (!(toProcess > 0 && cursor.hasNext())) {
          queue
        } else {
          val newQueue = queue.enqueue(cursor.next())
          if (queueLength >= toDrop) {
            val (a, nextQueue) = newQueue.dequeue
            buffer.append(a)
            queueLoop(nextQueue, toProcess - 1)
          } else {
            queueLength += 1
            queueLoop(newQueue, toProcess - 1)
          }
        }
      }
      val finalQueue = queueLoop(queue, limit)

      val next: F[Iterant[F, A]] = if (cursor.hasNext()) F.pure(ref) else rest
      NextBatch(Batch.fromSeq(buffer), next.map(loop(toDrop, queueLength, finalQueue)), stop)
    }

    def finalCursor(toDrop: Int, length: Int, queue: Queue[A]): Iterant[F, A] = {
      var queueLength = length
      val buffer = mutable.Buffer[A]()

      @tailrec
      def queueLoop(queue: Queue[A]): Queue[A] = {
        if (queueLength <= toDrop) {
          queue
        } else {
            val (a, nextQueue) = queue.dequeue
            buffer.append(a)
            queueLength -= 1
          queueLoop(nextQueue)
        }
      }
      queueLoop(queue)

      val cursor = BatchCursor.fromSeq(buffer)
      NextCursor(cursor, F.pure(Halt(None)), F.unit)
    }

    def loop(toDrop: Int, length: Int, queue: Queue[A])
            (source: Iterant[F, A]): Iterant[F, A] = {
      try if (toDrop <= 0) source else source match {
        case Next(elem, rest, stop) =>
          val newQueue = queue.enqueue(elem)
          if (length >= toDrop){
            val (next, nextQueue) = newQueue.dequeue
            Next(next, rest.map(loop(toDrop, length, nextQueue)), stop)
          }
          else Suspend(rest.map(loop(toDrop, length + 1, newQueue)), stop)
        case ref@NextCursor(_, _, _) =>
          processCursor(toDrop, length, queue, ref)
        case NextBatch(batch, rest, stop) =>
          processCursor(toDrop, length, queue, NextCursor(batch.cursor(), rest, stop))
        case Suspend(rest, stop) =>
          Suspend(rest.map(loop(toDrop, length, queue)), stop)
        case Last(item) =>
          finalCursor(toDrop, length + 1, queue.enqueue(item))
        case Halt(None) =>
          finalCursor(toDrop, length, queue)
        case halt@Halt(Some(_)) =>
          halt
      } catch {
        case ex if NonFatal(ex) =>
          val stop = source.earlyStop
          Suspend(stop.map(_ => Halt(Some(ex))), stop)
      }
    }

    // Suspending execution, because pushing into our queue
    // is side-effecting
    Suspend(F.delay(loop(n, 0, Queue.empty[A])(source)), source.earlyStop)
  }
}
```